org.osgi.test.common/src/main/java/org/osgi/test/common/annotation/ServicePropertyConverter.java

```
@@ -0,0 +1,183 @@
+		for (ServiceProperty entry : entrys) {
+			dictionary.put(entry.key(), toValue(entry));
+		}
+		return dictionary;
+	}
+
+	private static Object toValue(ServiceProperty entry) {
+
+		boolean primitive = entry.type()
+			.equals(Type.PrimitiveArray);
+		Object array = createArray(entry.scalar(), primitive, entry.value().length);
+		int i = 0;
+		for (String v : entry.value()) {
+			Object val = null;
+
+			if (v != null) {
+				switch (entry.scalar()) {
+					case Boolean :
+						Boolean booleanValue = Boolean.valueOf(v);
+						val = primitive ? booleanValue.booleanValue() : booleanValue;
+						break;
+
+					case Byte :
+						Byte byteVal = Byte.valueOf(v);
+						val = primitive ? byteVal.byteValue() : byteVal;
+						break;
+
+					case Character :
+						char charVal = v.charAt(0);
+						val = primitive ? charVal : new Character(charVal);
+						break;
+
+					case Double :
+						Double doubleVal = Double.valueOf(v);
+						val = primitive ? doubleVal.doubleValue() : doubleVal;
+						break;
+
+					case Float :
+						Float floatVal = Float.valueOf(v);
+						val = primitive ? floatVal.floatValue() : floatVal;
+						break;
+
+					case Integer :
+						Integer integerVal = Integer.valueOf(v);
+						val = primitive ? integerVal.intValue() : integerVal;
+						break;
+
+					case Long :
+						Long longVal = Long.valueOf(v);
+						val = primitive ? longVal.longValue() : longVal;
+						break;
+
+					case Short :
+						Short shortVal = Short.valueOf(v);
+						val = primitive ? shortVal.shortValue() : shortVal;
+						break;
+
+					case String :
+						val = v;
+						break;
+				}
+			}
+
+			if (Type.Scalar.equals(entry.type())) {
+				return val;
+			} else {
+				Array.set(array, i++, val);
+			}
+		}
+		if (entry.type()
+			.equals(Type.Array)
+			|| entry.type()
+				.equals(Type.PrimitiveArray)) {
+			return array;
+		} else if (entry.type()
+			.equals(Type.Collection)) {
+
+			return Arrays.asList((Object[]) array);
+		}
+
+		return null;
```

Returning null here guarantees an NPE in the Hashtable.put call above. If this is really a conversion error, you should probably throw a detailed exception here.

---

ok